### ISSUE TITLE
exception during LogManager.Instance.ReconfigExistingLoggers(true);

### ISSUE THREAD

**Issue by tvogel-nid** [NLog/NLog#4975]
Hi! Thanks for reporting this bug! 

Please keep / fill in the relevant info from this template so that we can help you as best as possible.

**NLog version**: 5.0.1

**Platform**:  .NET6

**Current NLog config** (xml or C#, if relevant)

```json
{
  "NLog": {
    "autoReload": true,
    "throwConfigExceptions": true,
    "internalLogLevel": "Warn",
    "internalLogFile": "${tempdir}/internal-nlog.txt",
    "extensions": [
      { "assembly": "NLog.Extensions.Logging" }
    ],
    "variables": {
      "var_datadir": "${tempdir}${dir-separator}niVCL${dir-separator}data${dir-separator}",
      "var_logdir": "${tempdir}${dir-separator}niVCL${dir-separator}",
      "NormalLayout": "${longdate} | ${pad:padding=5:inner=${level:uppercase=true}} | [threadId=${threadid}] [processId=${processid}] ${logger:shortName=true} | ${message} ${onexception:${newline}EXCEPTION\\: ${exception:format=ToString}}",
      "VerboseLayout": "${NormalLayout} | (${callsite:includeNamespace=false:className=true:methodName=true})"
    },
    "default-wrapper": {
      "type": "AsyncWrapper",
      "overflowAction": "Block"
    },
    "targets": {
      "log2file": {
        "type": "File",
        "fileName": "${var_logdir}${dir-separator}nlog-${shortdate}.log",
        "layout": "${NormalLayout}"
      },
      "log2console": {
        "type": "ColoredConsole",
        "layout": "${NormalLayout}"
      },
      "log2debugger": {
        "type": "DebugSystem",
        "layout": "${VerboseLayout}"
      }
    },
    "rules": [
      {
        "logger": "VectorErrorCorrectionModel*",
        "minLevel": "Info",
        "writeTo": "log2file, log2console, log2debugger",
        "final": true
      }
    ]
  }
}

```

Our project uses NLog in an probably unusual way.
We create a new logger and a new rule for it for each instance of a class.
New instances of the classes are created externally from unmanaged memory so there is no real way to exit clearly with LogManager.ShutDown() or to just dispose logger.
Over time we can see that the number of WeakReferences increases(seems to be the LoggerCache) so we tried to delete the created rule and then call LogManager.Instance.ReconfigExistingLoggers(true); in the destructor of the classes in the hope that these references are cleared and the GC can handle it.
What we can this is the exception below, so it seems that while the LogManager.Instance.ReconfigExistingLoggers(true); is running new logger are created from external calls and therefore the allocated Array is too small.
So it seems LogManager.Instance.ReconfigExistingLoggers(true); is not threadsafe even if I've found an older issue targeting exactly making it safe.

- What is the current result?
Exception occurs:
```
Message: 
System.ArgumentException : Destination array was not long enough. Check the destination index, length, and the array's lower bounds. (Parameter 'destinationArray')

  Stack Trace: 
Array.Copy(Array sourceArray, Int32 sourceIndex, Array destinationArray, Int32 destinationIndex, Int32 length, Boolean reliable)
Array.Copy(Array sourceArray, Int32 sourceIndex, Array destinationArray, Int32 destinationIndex, Int32 length)
List`1.CopyTo(T[] array, Int32 arrayIndex)
List`1.ctor(IEnumerable`1 collection)
Enumerable.ToList[TSource](IEnumerable`1 source)
LoggingConfiguration.GetLoggingRulesThreadSafe()
LoggingConfiguration.FindRuleByName(String ruleName)
VCLLogManager.FindRuleByName(String ruleName) line 87
VCLLogManager.CreateOrReconfigureRule(String ruleName, LogLevel minimumLogLevel) line 100
NDataLog.ConfigureDynamicRule(LogLevel minimumLogLevel) line 374
```
![image](https://user-images.githubusercontent.com/45204640/174101571-f09b69a8-077e-47ba-ae58-63b83682e74f.png)

- What is the expected result?
- Did you checked the [Internal log](https://github.com/NLog/NLog/wiki/Internal-Logging)?
- Please post full exception details (message, stacktrace, inner exceptions)
- Are there any workarounds? yes/no
- Is there a version in which it did work? No
- Can you help us by writing an unit test?

**Comment by tvogel-nid** [comment]
you might be right with your assumption although we use this approach for quite a while and just after trying to get rid of the increasing memory consumption via the ReconfigExistingLoggers(true) we see this exceptions during our unit tests.

**Comment by tvogel-nid** [comment]
well, I think you are right, if I look into the code it seems we fetch the Configuration, create an new Rule object and add it to the Configuration via `Configuration.LoggingRules.Add(rule)`
I think this is done this way as it seems there is no way to add a named rule via AddRule which prevents us to find the rule later via RemoveRuleByName, so if there is a save way to add a named rule this might be a way out

**Comment by snakefoot** [comment]
One way it is to ensure that only one application-thread is modfying the `LogManager.Configuration.LoggingRules` along with calling `ReconfigExistingLoggers(true) `

But I guess `NLog.LogManager.Configuration` could also get a new method called `AddRule(LoggingRule rule)`. Alternative the `LogManager.Configuration.LoggingRules` could be made threadsafe.

**Comment by snakefoot** [comment]
Resolved by #4976 and new `AddRule` releaed with https://www.nuget.org/packages/NLog/5.0.2
